util/src/Pos/Util/Wlog.hs

```diff
@@ -1,23 +1,288 @@
 import           System.Wlog.LogHandler (LogHandlerTag (HandlerFilelike))
+
+--import qualified Pos.Util.Log as Log
+
+import           Universum
+
+import           Control.Monad.Morph (MFunctor (..))
+import qualified Control.Monad.State.Lazy as StateLazy
+import           Data.Sequence ((|>))
+import           Data.Text (Text)
+
+{-
+  attempt for reducing complexity:
+
+  1. count call sites per function
+  2. comment out 0 usages
+  3. group by "common usage pattern"
+      * just logging functions
+      * setup logging
+      * logging configuration
+      * pure logger?
+      * only 'CanLog' imported
+      * CanLog and WithLogger imported => redundant
+      * CanLog and HasLoggerName imported => replace with 'WithLogger'
+-}
+
+{- compatibility -}
+
+-- setupLogging will setup global shared logging state (MVar)
+-- usingLoggerName launches an action (of 'LoggerNameBox m a' ~ 'ReaderT LoggerName m a')
+-- LoggerNameBox is an instance of CanLog, HasLoggerName
+-- our functions are running in this context: 'type WithLogger m = (CanLog m, HasLoggerName m)'
+
+-- this fails with 'MonadIO' redundant constraint
+-- type WithLogger m = (CanLog m, HasLoggerName m, Log.LogContext m)
+
+{-
+class GHC.Base.Monad m => CanLog (m :: * -> *) where
+  dispatchMessage :: LoggerName
+                     -> Severity -> Data.Text.Internal.Text -> m ()
+  default dispatchMessage :: (Control.Monad.Trans.Class.MonadTrans t,
+                              t n ~ m, CanLog n) =>
+                             LoggerName -> Severity -> Data.Text.Internal.Text -> m ()
+  -- Defined in ‘System.Wlog.CanLog’
+instance GHC.Base.Monad m => CanLog (NamedPureLogger m)
+  -- Defined in ‘System.Wlog.PureLogging’
+instance CanLog m => CanLog (LoggerNameBox m)
+  -- Defined in ‘System.Wlog.CanLog’
+-}
+class Monad m => CanLog m where
+    dispatchMessage :: LoggerName -> Severity -> Text -> m ()
+
+    default dispatchMessage :: (MonadTrans t, t n ~ m, CanLog n)
+                            => LoggerName
+                            -> Severity
+                            -> Text
+                            -> m ()
+    dispatchMessage name sev t = lift $ dispatchMessage name sev t
+
+instance CanLog m => CanLog (LoggerNameBox m)
+instance CanLog m => CanLog (ReaderT r m)
+instance CanLog m => CanLog (StateT s m)
+instance CanLog m => CanLog (StateLazy.StateT s m)
+instance CanLog m => CanLog (ExceptT s m)
+instance CanLog IO where
+    dispatchMessage ln _ t = putStrLn $ "[" ++ (show ln) ++ "] " ++ (show t)
+
+type WithLogger m = (CanLog m, HasLoggerName m)
+
+{- dummies for now -}
+logDebug, logError, logInfo, logNotice, logWarning
+  :: (CanLog m) => Text -> m ()
+logDebug _ = return ()
+logInfo _ = return ()
+logError _ = return ()
+logNotice _ = return ()
+logWarning _ = return ()
+
+logMessage :: (CanLog m) => Severity -> Text -> m ()
+logMessage _ _ = return ()
+
+{-
+type role LoggerNameBox representational nominal
```

Comment:
Is this still needed?

Reply:
this is actually the central logging function which offloads the message with a severity to `CanLog` using `dispatchMessage`, and this then passes the message to `katip`.